server/src/main/java/com/thoughtworks/go/server/service/RulesService.java

```diff
@@ -44,13 +48,42 @@ public RulesService(GoConfigService goConfigService) {
     public boolean validateSecretConfigReferences(ScmMaterial scmMaterial) {
         List<CaseInsensitiveString> pipelines = goConfigService.pipelinesWithMaterial(scmMaterial.getFingerprint());
         SecretParams secretParams = scmMaterial.getSecretParams();
-        pipelines.forEach(pipeline -> {
-            PipelineConfigs group = goConfigService.findGroupByPipeline(pipeline);
-            String errorMessagePrefix = format("Material with url: '%s' in Pipeline: '%s' and Pipeline Group:", scmMaterial.getUriForDisplay(), pipeline);
-            validateSecretConfigReferences(secretParams, group.getClass(), group.getGroup(), errorMessagePrefix);
-        });
 
+        List<String> missingSecretConfigs = secretParams.stream()
+                .filter(secretParam -> goConfigService.cruiseConfig().getSecretConfigs().find(secretParam.getSecretConfigId()) == null)
+                .map(SecretParam::getSecretConfigId)
+                .collect(Collectors.toList());
+
+        if (!missingSecretConfigs.isEmpty()) {
+            throwSecretConfigNotFound("ScmMaterial", scmMaterial.getUriForDisplay(), String.join(", ", missingSecretConfigs));
+        }
+
+        HashMap<CaseInsensitiveString, List<String>> pipelinesWithViolatedSecretConfigIds = new HashMap<>();
+        secretParams.stream()
+                .map(secretParam -> goConfigService.cruiseConfig().getSecretConfigs().find(secretParam.getSecretConfigId()))
+                .forEach(secretConfig -> {
+                    pipelines.forEach(pipelineName -> {
+                        PipelineConfigs group = goConfigService.findGroupByPipeline(pipelineName);
+                        if (!secretConfig.canRefer(group.getClass(), group.getGroup())) {
+                            if (!pipelinesWithViolatedSecretConfigIds.containsKey(pipelineName)) {
+                                pipelinesWithViolatedSecretConfigIds.put(pipelineName, new ArrayList<>());
+                            }
+                            LOGGER.error("[Material Update] Failure", format("Pipeline '%s' does not have permission to refer to secrets using SecretConfig: '%s'", pipelineName.toString(), secretConfig.getId()));
+                            pipelinesWithViolatedSecretConfigIds.get(pipelineName).add(secretConfig.getId());
+
+                        }
+                    });
+                });
+
+        if (pipelinesWithViolatedSecretConfigIds.size() == pipelines.size()) {
+            String secretConfigId = secretParams.get(0).getSecretConfigId();
+            String errorMessage = format("Material with url: '%s' does not have permissions to refer to secret config with id: '%s'. The pipelines which use this material are: '%s'", scmMaterial.getUriForDisplay(), secretConfigId, pipelinesWithViolatedSecretConfigIds.keySet().toString());
```

Comment:
Since the material password is not part of the fingerprint. Different pipeline with the same material can have a password configured using a different secret config ID. In that case, the message will always print the first secret config ID and might be misleading.

What about the following the message?
```
Material with URL: 'some-url' does not have permissions to refer to secret config 'aws' in pipelines "prod", "pre-prod" and secret config 'file' in pipelines "test", "build"
```
/cc @maheshp